### PR TITLE
Avoid editable installs on CI

### LINKED ISSUE #8854: Invoking tests against editable installs have been masking bugs
We've been invoking tests on CI against an editable install of pylint for a long time. Doing so has the potential to mask issues. See [blog post](https://blog.ganssle.io/articles/2019/08/test-as-installed.html).

We now have evidence of some of those issues from the [failures](https://github.com/pylint-dev/pylint/actions/runs/5556300913/jobs/10148631043) in the PR testing against Python 3.12, which for unrelated reasons, forced us to upgrade `setuptools`. Without using the [legacy behavior](https://setuptools.pypa.io/en/latest/userguide/development_mode.html#legacy-behavior), bugs became evident.

One of the most minimal bugs is this one:

https://github.com/pylint-dev/pylint/blob/9f2de9123b579fadd512e31e9b68c7eb40a9c62d/tests/lint/test_pylinter.py#L39-L41

That mock shouldn't work, since the real location of the `PyLinter` class is at `pylint.lint.pylinter.PyLinter`. It was only working because the `pylint.lint` module imports `PyLinter`, but AFIAU there's no reason to expect that to work when testing the installation of the `pylint` package, which won't install `pylint.lint` (whereas an editable install will, assuming you're invoking tests from `pylint`.

That's just a bug in a test, but there are functional bugs, too, such as the moved check in #8702 potentially crashing:

```python
File "/Users/<user>/pylint/pylint/config/config_initialization.py", line 114, in _config_initialization
    for exc_name in linter.config.overgeneral_exceptions:
AttributeError: 'Namespace' object has no attribute 'overgeneral_exceptions'
```

***

I've found a workaround for the 3.12 branch to keep that one using legacy behavior, but suggesting we also look at this fixing this for the next pylint alpha.

### PR DESCRIPTION
**Pull request by jacobtylerwalls**
## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :bug: Bug fix          |

## Description
Closes #8854

### CODE REVIEW

File: requirements_test_min.txt [hunk 1]
```diff
@@ -1,4 +1,4 @@
--e .[testutils,spelling]
+.[testutils,spelling]
```
Comment (Pierre-Sassoulas) [inline requirements_test_min.txt:1]: It's convenient to be able to do editable install when developing locally though, could we keep this one maybe ?

Comment (jacobtylerwalls) [reply requirements_test_min.txt:1]: Is updating the contributor docs to specify that an okay compromise? (I'm not used to expecting requirements files to take care of that for me.)

Comment (jacobtylerwalls) [reply requirements_test_min.txt:1]: CI installs from this file, so we'd have to do a bigger refactor otherwise.

Comment (Pierre-Sassoulas) [reply requirements_test_min.txt:1]: You're right, fait enough

**Review by Pierre-Sassoulas — APPROVED** [review]

**Comment by jacobtylerwalls** [comment]
d441a07d19b44054a50cd40301f28d96dd0f5b0a will likely cause devs with editable installs to invoke pytest with `python -m pytest` instead of `pytest` but that's a small cost to pay for starting to do this more correctly...

**Comment by jacobtylerwalls** [comment]
However, I do so and I get dozens of failures. I'd appreciate eyes on this; the current situation is a mess.

**Comment by jacobtylerwalls** [comment]
I'm less concerned now. These are all passing locally, save for some longstanding trouble with the `forgotten-debug-statement` test.
```sh
cd pylint
pip install .
pytest
python3 -m pytest
```
This fails, but status quo...
```sh
cd ..
pytest pylint # fails, but status quo...
```

File: tests/test_check_parallel.py [hunk 2]
```diff
@@ -269,9 +269,10 @@ def test_linter_with_unpickleable_plugins_is_pickleable(self) -> None:
         linter.load_plugin_modules(["pylint.extensions.overlapping_exceptions"])
         try:
             dill.dumps(linter)
-            raise AssertionError(
-                "Plugins loaded were pickle-safe! This test needs altering"
-            )
+            # TODO: 3.0: Fix this test by raising this assertion again
```
Comment (Pierre-Sassoulas) [inline tests/test_check_parallel.py:272]: Having to remove the pickle safe tests is a little frightening.

File: doc/development_guide/contributor_guide/tests/install.rst [hunk 3]
```diff
@@ -20,6 +20,7 @@ environment for testing, open a terminal and run::
     python3 -m venv venv
     source venv/bin/activate
     pip install -r requirements_test_min.txt
+    pip install -e .
```
Comment (Pierre-Sassoulas) [inline doc/development_guide/contributor_guide/tests/install.rst:23]: 👍

**Review by Pierre-Sassoulas — COMMENTED** [review]
Yeah, we're just going to launch pytest from the root of the git repo, I've never once in the last 10 years ``cded`` away from the root of a repo when working on the code (I do ``cd doc`` and launch the doc tests from there from time to time). Maybe we just need to document it.  I don't think I understand the problem we're facing here though. (Also, was ``tox`` protecting us from the issues that could arise because it's installing/isolating the package its testing ?)

**Comment by jacobtylerwalls** [comment]
I appreciate your giving this some consideration!

> Maybe we just need to document it.

I'm fine leaving this small decision for another day; was something I just happened to notice.

> I don't think I understand the problem we're facing here though.

We're testing against the editable install and with `/pylint` appended to sys.path, neither of which will be the case for our users.

>  (Also, was tox protecting us from the issues that could arise because it's installing/isolating the package its testing ?)

Yes, `tox` is probably saving our bacon with respect to one of those two issues, but perhaps not both, which is why the pickle test needed altering. (And why I think it's okay to go ahead with this change.)

> I'm less concerned now.

My bug report suggested there were crashes lurking behind all this. I wonder if I just had outdated versions of setuptools. Let's just see if anybody reports bugs in the beta.

> Having to remove the pickle safe tests is a little frightening.

I get that, but we're just returning things to the state of play before the assertion was commented out without investigation. If a test is not providing real value, better to know that than to suggest everything is A-OK because we decided to use the trick of using an editable install.

**Review by Pierre-Sassoulas — APPROVED** [review]

**Comment by jacobtylerwalls** [comment]
I'm a little concerned about codecov whiffing, but maybe we can merge and see what happens.

**Comment by jacobtylerwalls** [comment]
Looking into the doc and codecov failures on main...

**Pull request merged**